# Performing EDA

In [19]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

In [15]:
# Loading dataset
df = pd.read_csv(
    "../data/exoplanets_raw.csv",
    comment="#",         # drop any line beginning with “#”
    engine="python",     # forgiving parser
    on_bad_lines="skip"  # skip rows that still don’t line up
)

print("Shape:", df.shape)
print("Columns:", df.columns[:10], "…")
df.head()


Shape: (5917, 168)
Columns: Index(['rowid', 'pl_name', 'hostname', 'pl_letter', 'hd_name', 'hip_name',
       'tic_id', 'gaia_id', 'sy_snum', 'sy_pnum'],
      dtype='object') …


rowid     pl_name  hostname pl_letter    hd_name    hip_name  \
0      1    11 Com b    11 Com         b  HD 107383   HIP 60202   
1      2    11 UMi b    11 UMi         b  HD 136726   HIP 74793   
2      3    14 And b    14 And         b  HD 221345  HIP 116076   
3      4    14 Her b    14 Her         b  HD 145675   HIP 79248   
4      5  16 Cyg B b  16 Cyg B         b  HD 186427   HIP 96901   

          tic_id                       gaia_id  sy_snum  sy_pnum  ...  \
0   TIC 72437047  Gaia DR2 3946945413106333696        2        1  ...   
1  TIC 230061010  Gaia DR2 1696798367260229376        1        1  ...   
2  TIC 333225860  Gaia DR2 1920113512486282240        1        1  ...   
3  TIC 219483057  Gaia DR2 1385293808145621504        1        2  ...   
4   TIC 27533327  Gaia DR2 2135550755683407232        3        1  ...   

                                     sy_tmag_reflink  sy_kepmag  \
0  <a refstr=STASSUN_ET_AL__2019 href=https://ui....        NaN   
1  <a refstr=STASSUN_ET_AL__2019 href=https://ui....        NaN   
2  <a refstr=STASSUN_ET_AL__2019 href=https://ui....        NaN   
3  <a refstr=STASSUN_ET_AL__2019 href=https://ui....        NaN   
4  <a refstr=STASSUN_ET_AL__2019 href=https://ui....      6.095   

                                   sy_kepmag_reflink  pl_nnotes st_nphot  \
0  <a refstr=STASSUN_ET_AL__2019 href=https://ui....        2.0      1.0   
1  <a refstr=STASSUN_ET_AL__2019 href=https://ui....        0.0      1.0   
2  <a refstr=STASSUN_ET_AL__2019 href=https://ui....        0.0      1.0   
3  <a refstr=STASSUN_ET_AL__2019 href=https://ui....        0.0      1.0   
4  <a refstr=STASSUN_ET_AL__2019 href=https://ui....        5.0      1.0   

  st_nrvc st_nspec pl_nespec pl_ntranspec pl_ndispec  
0     2.0      0.0       0.0          0.0        0.0  
1     1.0      0.0       0.0          0.0        0.0  
2     1.0      0.0       0.0          0.0        0.0  
3     4.0      1.0       0.0          0.0        0.0  
4     4.0      3.0       0.0          0.0        0.0  

[5 rows x 168 columns]

In [17]:
# 1) Overview of dtypes and non-null counts
df.info(verbose=True, show_counts=True)

# 2) Percent missing per column
missing_pct = df.isna().mean().sort_values(ascending=False) * 100
missing_pct.head(10)  # top 10 most-missing columns

# 3) A quick look at our key numeric features
key_feats = [
    "pl_rade",    # planet radius (Earth radii)
    "pl_bmasse",  # planet mass (Earth masses)
    "pl_orbeccen",
    "pl_insol",   # insolation (Earth flux)
    "pl_eqt",     # equilibrium temperature
    "pl_orbsmax", # semi-major axis (AU)
    "st_teff",    # stellar effective temp (K)
    "st_rad",     # stellar radius (Solar radii)
    "st_mass"     # stellar mass (Solar masses)
]
df[key_feats].describe().T


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5917 entries, 0 to 5916
Data columns (total 168 columns):
 #    Column                Non-Null Count  Dtype  
---   ------                --------------  -----  
 0    rowid                 5917 non-null   int64  
 1    pl_name               5917 non-null   object 
 2    hostname              5917 non-null   object 
 3    pl_letter             5917 non-null   object 
 4    hd_name               1007 non-null   object 
 5    hip_name              1081 non-null   object 
 6    tic_id                5656 non-null   object 
 7    gaia_id               5606 non-null   object 
 8    sy_snum               5917 non-null   int64  
 9    sy_pnum               5917 non-null   int64  
 10   sy_mnum               5917 non-null   int64  
 11   cb_flag               5917 non-null   int64  
 12   discoverymethod       5917 non-null   object 
 13   disc_year             5917 non-null   int64  
 14   disc_refname          5917 non-null   object 
 15   di

count         mean          std       min          25%  \
pl_rade      5890.0     5.752880     5.327163    0.3098     1.800441   
pl_bmasse    5881.0   382.978577  1094.608200    0.0200     4.100000   
pl_orbeccen  5048.0     0.077583     0.151826   -0.0110     0.000000   
pl_insol     4122.0   419.977100  1336.094476    0.0003    23.837000   
pl_eqt       4394.0   912.442167   463.763696   34.0000   568.000000   
pl_orbsmax   5621.0    15.732620   355.178932    0.0044     0.052700   
st_teff      5642.0  5411.476554  1765.576997  415.0000  4919.250000   
st_rad       5616.0     1.517643     3.997636    0.0115     0.771000   
st_mass      5892.0     0.941749     0.422867    0.0094     0.777000   

                   50%          75%           max  
pl_rade         2.8100    11.900000     77.342100  
pl_bmasse       8.9600   174.487795   9344.155166  
pl_orbeccen     0.0000     0.090000      0.950000  
pl_insol       96.9405   357.064575  44900.000000  
pl_eqt        818.5000  1159.000000   4050.000000  
pl_orbsmax      0.1027     0.302000  19000.000000  
st_teff      5550.0000  5900.000000  57000.000000  
st_rad          0.9500     1.242750     88.475000  
st_mass         0.9400     1.090000     10.940000

In [18]:
# 1) Read the raw data (if not already in memory)
df = pd.read_csv("../data/exoplanets_raw.csv", comment="#", engine="python", on_bad_lines="skip")

# 2) Define the core features and label
features = [
    "pl_rade", "pl_bmasse", "pl_orbsmax",
    "pl_orbeccen", "pl_insol",
    "st_teff", "st_rad", "st_mass"
]
df["habitable"] = df["pl_insol"].between(0.3, 1.5).astype(int)

# 3) Subset and drop any rows with missing values in these columns
model_df = df[features + ["habitable"]].dropna()

# 4) Save to a new CSV
out_path = "../data/exoplanets_processed.csv"
model_df.to_csv(out_path, index=False)
print(f"Processed dataset saved to {out_path} with shape {model_df.shape}")


Processed dataset saved to ../data/exoplanets_processed.csv with shape (3816, 9)


## Conclusion

In this notebook, we performed the initial data‐preparation steps needed to turn the raw NASA Exoplanet Archive table into a clean, modeling-ready dataset. Here’s a summary of what we did:

1. **Loaded the raw CSV**  
   - Read in `exoplanets_raw.csv`, skipping the header comments and any malformed rows.  
   - Verified that we had 5,917 confirmed exoplanets and 168 columns of mixed metadata, measurements, and HTML reference links.

2. **Chose our core modeling features**  
   We selected eight physically meaningful, high-coverage columns that drive habitability predictions:

   | Feature        | Units                  | Why we picked it                                                             |
   |---------------:|-----------------------:|------------------------------------------------------------------------------|
   | **pl_rade**    | Earth radii            | Planet size: distinguishes rocky vs. gaseous worlds.                         |
   | **pl_bmasse**  | Earth masses           | Planet mass: with radius, infers density and surface conditions.             |
   | **pl_orbsmax** | Astronomical units     | Distance from host star: primary control on temperature and flux.            |
   | **pl_orbeccen**| (unitless)             | Orbital eccentricity: high values → extreme temperature swings.              |
   | **pl_insol**   | Earth flux             | Insolation: direct proxy for whether a planet lies in the “Goldilocks” zone. |
   | **st_teff**    | Kelvin                 | Stellar temperature: combined with radius defines the star’s luminosity.     |
   | **st_rad**     | Solar radii            | Stellar size: helps set the inner/outer bounds of the habitable zone.       |
   | **st_mass**    | Solar masses           | Stellar mass: correlates with luminosity and stellar lifetime.               |

3. **Created a binary habitability label**  
   ```python
   df["habitable"] = df["pl_insol"].between(0.3, 1.5).astype(int)
- 1 = “habitable”: receives between 0.3× and 1.5× Earth’s insolation

- 0 = “non-habitable”: outside that flux range (or missing)

4. **Filtered to complete cases**

- Subset the DataFrame to these nine columns (8 features + habitable)

- Dropped any row with missing values, leaving 3,816 planets with a full feature vector

5. **Saved the processed dataset**
   ```python
   model_df.to_csv("../data/exoplanets_processed.csv", index=False)
- Leaves the raw file untouched

- Produces a clean, small CSV that downstream scripts can load directly for training and evaluation

6. **Next Steps**

With exoplanets_processed.csv in hand, we will:

- Split into training and validation sets

- Normalize each feature (e.g. standard scaling)

- Build a PyTorch Dataset & DataLoader

- Define and train a feedforward neural network

- Evaluate with metrics like accuracy, precision, recall, and ROC-AUC